src/parser/mod.rs

```diff
@@ -103,6 +109,67 @@ pub struct KeyMetadata {
     pub size: u32,
 }
 
+/// A trait for parsing multiple types of values to memcached values.
+pub trait ToMemcachedValue {
+    /// Returns the length of the value in bytes.
+    fn length(&self) -> usize;
+    /// Writes the value to a writer.
+    fn write_to<W: AsyncWriteExt + Unpin>(
+        &self,
+        writer: &mut W,
+    ) -> impl Future<Output = Result<(), crate::Error>>;
+}
+
+impl ToMemcachedValue for &[u8] {
```

Comment:
Can we do this so that we can continue to support all types that we did previously
```suggestion
impl<T: AsRef<[u8]>> ToMemcachedValue for T {
```

Reply:
When going down this route I end up having to create a standalone wrapper struct for u64 inputs which sort of defeats the intention behind this change, since it ends up making it more awkward to use a `uint` value.

I've included `String` and `&String` implementations for `ToMemcachedValue` which I think covers everything from `AsRef` that we would realistically see as input for the `set` and `add` functions here.  For reference, this [rust-memcache crate](https://github.com/aisk/rust-memcache) is where we picked up this pattern from and they have implemented on [the same set of types](https://github.com/aisk/rust-memcache/blob/2e952856dd9c2b45c5f0d87b6a135471e94627a5/src/value.rs#L12) that I have updated this PR to include.